libs/utils/src/http/endpoint.rs

```diff
@@ -148,26 +147,129 @@ impl Drop for RequestCancelled {
 }
 
 async fn prometheus_metrics_handler(_req: Request<Body>) -> Result<Response<Body>, ApiError> {
+    use bytes::{Bytes, BytesMut};
+    use std::io::Write as _;
+    use tokio::sync::mpsc;
+    use tokio_stream::wrappers::ReceiverStream;
+
     SERVE_METRICS_COUNT.inc();
 
-    let mut buffer = vec![];
-    let encoder = TextEncoder::new();
+    /// An [`std::io::Write`] implementation on top of a channel sending [`bytes::Bytes`] chunks.
+    struct ChannelWriter {
+        buffer: bytes::BytesMut,
+        tx: mpsc::Sender<std::io::Result<bytes::Bytes>>,
+        written: usize,
+    }
 
-    let metrics = tokio::task::spawn_blocking(move || {
-        // Currently we take a lot of mutexes while collecting metrics, so it's
-        // better to spawn a blocking task to avoid blocking the event loop.
-        metrics::gather()
-    })
-    .await
-    .map_err(|e: JoinError| ApiError::InternalServerError(e.into()))?;
-    encoder.encode(&metrics, &mut buffer).unwrap();
+    impl ChannelWriter {
+        fn new(buf_len: usize, tx: mpsc::Sender<std::io::Result<Bytes>>) -> Self {
+            ChannelWriter {
+                buffer: BytesMut::with_capacity(buf_len),
+                tx,
+                written: 0,
+            }
+        }
+
+        fn flush0(&mut self) -> std::io::Result<usize> {
+            let n = self.buffer.len();
+            if n == 0 {
+                return Ok(0);
+            }
+
+            tracing::trace!(n, "flushing");
+            let ready = self.buffer.split().freeze();
+
+            // not ideal to call from blocking code to block_on, but we are sure that this
+            // operation does not spawn_blocking other tasks
+            let res: Result<(), ()> = tokio::runtime::Handle::current().block_on(async {
+                self.tx.send(Ok(ready)).await.map_err(|_| ())?;
+
+                // throttle sending to allow reuse of our buffer in `write`.
+                self.tx.reserve().await.map_err(|_| ())?;
+
+                // now the response task has picked up the buffer and hopefully started
+                // sending it to the client.
+                Ok(())
+            });
+            if res.is_err() {
+                return Err(std::io::ErrorKind::BrokenPipe.into());
+            }
+            self.written += n;
+            Ok(n)
+        }
+
+        fn flushed_bytes(&self) -> usize {
+            self.written
+        }
+    }
+
+    impl std::io::Write for ChannelWriter {
+        fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+            let remaining = self.buffer.capacity() - self.buffer.len();
+
+            let out_of_space = remaining < buf.len();
+
+            if out_of_space {
+                self.flush0()?;
+            }
```

Comment:
This is still a real problem. It might cause a lot of allocations. It would be helpful to ... Only do the half of the buffer at a time.

Perhaps I figured it just out; no code changes here, but split the buffer at half *after* allocating it. Then it will grow up to or near the "half", and extend_from_slice will be taking up the rest of the allocation.

Reply:
Well, caveats remain even with this. There are cases where the prometheus does do small writes, but perhaps we can asusme this is fine.

Reply:
At least with fast localhost connection this works, but probably breaks down with slower.

```
2023-07-21T13:14:55.435907Z  INFO request{method=GET path=/metrics request_id=4e6cd653-99f7-409f-bd64-19f31cda7426}:blocking: flushing n=65563
2023-07-21T13:14:55.436134Z  INFO request{method=GET path=/metrics request_id=4e6cd653-99f7-409f-bd64-19f31cda7426}:blocking: flushing n=65482
2023-07-21T13:14:55.436321Z  INFO request{method=GET path=/metrics request_id=4e6cd653-99f7-409f-bd64-19f31cda7426}:blocking: flushing n=65491
2023-07-21T13:14:55.436504Z  INFO request{method=GET path=/metrics request_id=4e6cd653-99f7-409f-bd64-19f31cda7426}:blocking: flushing n=65535
2023-07-21T13:14:55.436783Z  INFO request{method=GET path=/metrics request_id=4e6cd653-99f7-409f-bd64-19f31cda7426}:blocking: flushing n=65508
2023-07-21T13:14:55.436963Z  INFO request{method=GET path=/metrics request_id=4e6cd653-99f7-409f-bd64-19f31cda7426}:blocking: flushing n=65530
2023-07-21T13:14:55.437146Z  INFO request{method=GET path=/metrics request_id=4e6cd653-99f7-409f-bd64-19f31cda7426}:blocking: flushing n=65541
2023-07-21T13:14:55.437348Z  INFO request{method=GET path=/metrics request_id=4e6cd653-99f7-409f-bd64-19f31cda7426}:blocking: flushing n=65528
2023-07-21T13:14:55.437537Z  INFO request{method=GET path=/metrics request_id=4e6cd653-99f7-409f-bd64-19f31cda7426}:blocking: flushing n=65536
2023-07-21T13:14:55.437786Z  INFO request{method=GET path=/metrics request_id=4e6cd653-99f7-409f-bd64-19f31cda7426}:blocking: flushing n=57983
2023-07-21T13:14:55.437816Z  INFO request{method=GET path=/metrics request_id=4e6cd653-99f7-409f-bd64-19f31cda7426}:blocking: responded /metrics bytes=647697 elapsed_ms=2
```

Assuming good enough for now.

Reply:
(I forgot to return the full length from the write.)

Reply:
[1ccf599](https://github.com/neondatabase/neon/pull/4768/commits/1ccf599548937ca8d1e8784074c718a1d2034d29)